### PR TITLE
fix(stepfunctions-tasks): confusion between multiple ways to run a Lambda

### LINKED ISSUE #4801: aws-stepfunctions-tasks: Confusion between RunLambdaTask and InvokeFunction
There are now two similar-looking constructs in aws-stepfunctions-tasks pertaining to Lambda tasks, and it's not clear which one to use.

Users who are used to historical Step Functions patterns and want things to easily work should use `InvokeFunction`.  However, there are some corner cases where people want special behavior (e.g., asynchronous execution) who like to use `RunLambdaTask`.

If you mix these up, be prepared to spend a lot of time debugging.  @ccfife and I spent the afternoon scratching our heads to figure out, when we accidentally used `RunLambdaTask`, why the event payload being sent to our Lambda function Tasks was an empty event.

Let's do something to disambiguate these and make clear when a user should use one or the other.

### PR DESCRIPTION
**Pull request by shivlaks**
### Commit Message

fix(stepfunctions-tasks): confusion between multiple ways to run a Lambda

The `InvokeFunction` Step Functions task is being marked as deprecated. It represents the legacy way to represent Lambda functions in Step Functions 

The `RunLambda` task represents the recommended way to invoke Lambdas in Step Functions.
see: https://docs.aws.amazon.com/step-functions/latest/dg/connect-lambda.html

Examples in the README have been updated to use `RunLambdaTask`

Closes #4801

### Commit End

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-stepfunctions/README.md [hunk 1]
```diff
@@ -35,7 +35,7 @@ const submitLambda = new lambda.Function(this, 'SubmitLambda', { ... });
 const getStatusLambda = new lambda.Function(this, 'CheckLambda', { ... });
 
 const submitJob = new sfn.Task(this, 'Submit Job', {
-    task: new tasks.InvokeFunction(submitLambda),
+    task: new tasks.RunLambdaTask(submitLambda),
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-stepfunctions/README.md]: Given that a docs section below says the following:

```
Except `tasks.InvokeActivity`, the [service integration
pattern](https://docs.aws.amazon.com/step-functions/latest/dg/connect-to-resource.html)
(`integrationPattern`) are supposed to be given as parameter when customers want
to call integrated services within a Task state. The default value is `FIRE_AND_FORGET`.
```

Either the docs are wrong or these two calls cannot be equivalent.

The original one (`InvokeFunction`) would call a Lambda function, wait for it to finish and add the Lambda's return value to the state object.

If the default integration pattern is indeed `FIRE_AND_FORGET`, then this call will now call `Invoke`, add the [response object of `Invoke`](https://docs.aws.amazon.com/lambda/latest/dg/API_Invoke.html#API_Invoke_ResponseElements) to the state object, and have moved on to the next state by the time the Lambda function finishes.

By the way, I think that `FIRE_AND_FORGET` is a silly default. How many people will be writing StepFunctions state machines to fire off tasks one after the other and not care about the result? Versus how many are doing it to stitch together long-running operations where they DO care about the outcome? `SYNC` would make a much better default for nearly all cases, methinks.

Comment (shivlaks) [reply packages/@aws-cdk/aws-stepfunctions/README.md]: good point, i'll update the example.

I agree with the sentiment on the default - any thoughts there @wqzoww

**Comment by rix0rrr** [comment]
I think I also generally disagree with removing the class. There will be a lot of people we will be breaking, which we need to think of. Let's `@deprecate`, and remove from the docs and examples instead?

**Comment by nija-at** [comment]
> I think I also generally disagree with removing the class. There will be a lot of people we will be breaking, which we need to think of. Let's @deprecate, and remove from the docs and examples instead?

After a chat with @rix0rrr and some further though, he has convinced me that this is probably a more customer friendly way to changing this. 
However, we should make it clear that no new features or improvements should be expected on this.
We should also traverse all of our documentation and examples code (including aws-cdk-examples) to find and replace InvokeFunction with RunLambda.

**Comment by jogold** [comment]
Is the [handling of Lambda service exceptions](https://docs.aws.amazon.com/step-functions/latest/dg/bp-lambda-serviceexception.html) the same when using the `states` arn vs the full function arn? The documentation is really not clear about this...

**Comment by shivlaks** [comment]
> Is the [handling of Lambda service exceptions](https://docs.aws.amazon.com/step-functions/latest/dg/bp-lambda-serviceexception.html) the same when using the `states` arn vs the full function arn? Are Lambda errors reported in both cases? The documentation is really not clear about this...

good question, I'll follow up on these today

**Comment by shivlaks** [comment]
> > I think I also generally disagree with removing the class. There will be a lot of people we will be breaking, which we need to think of. Let's @deprecate, and remove from the docs and examples instead?
> 
> After a chat with @rix0rrr and some further though, he has convinced me that this is probably a more customer friendly way to changing this.
> However, we should make it clear that no new features or improvements should be expected on this.
> We should also traverse all of our documentation and examples code (including aws-cdk-examples) to find and replace InvokeFunction with RunLambda.

sure, this is where I started out with this change before our decision to remove it. I also think it's more friendly to deprecate it and encourage the preferred flow to allow users the option of using it. 

I like the suggestion to add some wording around future development and maintenance. Will make these changes.

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/invoke-function.ts [hunk 2]
```diff
@@ -22,6 +22,8 @@ export interface InvokeFunctionProps {
  * The Lambda function Arn is defined as Resource in the state machine definition.
  *
  * OUTPUT: the output of this task is the return value of the Lambda Function.
+ *
+ * @deprecated Use `RunLambdaTask`
```
Comment (nija-at) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/invoke-function.ts]: Can we also mark `InvokeFunctionProps` as deprecated in a similar way

**Comment by shivlaks** [comment]
> Is the [handling of Lambda service exceptions](https://docs.aws.amazon.com/step-functions/latest/dg/bp-lambda-serviceexception.html) the same when using the `states` arn vs the full function arn? Are Lambda errors reported in both cases? The documentation is really not clear about this...

good question, investigating!

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/invoke-function.ts (hunk 2 above)
Comment (shivlaks) [reply packages/@aws-cdk/aws-stepfunctions-tasks/lib/invoke-function.ts]: done

**Comment by shivlaks** [comment]
> > Is the [handling of Lambda service exceptions](https://docs.aws.amazon.com/step-functions/latest/dg/bp-lambda-serviceexception.html) the same when using the `states` arn vs the full function arn? Are Lambda errors reported in both cases? The documentation is really not clear about this...
> 
> good question, investigating!

@jogold 

Looked into this a bit further
* Lambda service exceptions are handled the same whether we define them through either mechanism
* Lambda Errors are reported in both cases

I also reached out to a member of the Step Functions team who verified that they should be the same

**Comment by jogold** [comment]
> > > Is the [handling of Lambda service exceptions](https://docs.aws.amazon.com/step-functions/latest/dg/bp-lambda-serviceexception.html) the same when using the `states` arn vs the full function arn? Are Lambda errors reported in both cases? The documentation is really not clear about this...
> > 
> > 
> > good question, investigating!
> 
> @jogold
> 
> Looked into this a bit further
> 
> * Lambda service exceptions are handled the same whether we define them through either mechanism
> * Lambda Errors are reported in both cases
> 
> I also reached out to a member of the Step Functions team who verified that they should be the same

Great! Thank you.

**Comment by shivlaks** [comment]
> LGTM.
> 
> It would be useful to customers if we can configure the [retry policy](https://docs.aws.amazon.com/step-functions/latest/dg/bp-lambda-serviceexception.html) and they will get this for free (with options to adjust).
> 
> This, however, is quite hard to implement in our current implementation. One more case to do this - #6715. If we changed the pattern to what is described here, retry policy for each type of task can be configured within it.

good point, I'll take a look into this next

**Pull request merged**